src/db_impl.cc

```diff
@@ -583,6 +583,145 @@ void TitanDBImpl::ReleaseSnapshot(const Snapshot* snapshot) {
   db_->ReleaseSnapshot(snapshot);
 }
 
+Status TitanDBImpl::DeleteFilesInRanges(ColumnFamilyHandle* column_family,
+                                        const RangePtr* ranges, size_t n,
+                                        bool include_end) {
+  TablePropertiesCollection props;
+  auto cfh = reinterpret_cast<ColumnFamilyHandleImpl*>(column_family);
+  auto cfd = cfh->cfd();
+  Version* version;
+
+  // Increment the ref count
+  {
+    InstrumentedMutexLock l(db_impl_->mutex());
+    version = cfd->current();
+    version->Ref();
+  }
+
+  auto* vstorage = version->storage_info();
+  for (size_t i = 0; i < n; i++) {
+    auto begin = ranges[i].start, end = ranges[i].limit;
+    // Get all the files within range except L0, cause `DeleteFilesInRanges`
+    // would not delete the files in L0.
+    for (int level = 1; level < vstorage->num_non_empty_levels(); level++) {
+      if (vstorage->LevelFiles(i).empty() ||
+          !vstorage->OverlapInLevel(i, begin, end)) {
+        continue;
+      }
+      std::vector<FileMetaData*> level_files;
+      InternalKey begin_storage, end_storage, *begin_key, *end_key;
+      if (begin == nullptr) {
+        begin_key = nullptr;
+      } else {
+        begin_storage.SetMinPossibleForUserKey(*begin);
+        begin_key = &begin_storage;
+      }
+      if (end == nullptr) {
+        end_key = nullptr;
+      } else {
+        end_storage.SetMaxPossibleForUserKey(*end);
+        end_key = &end_storage;
+      }
+
+      std::vector<FileMetaData*> files;
+      vstorage->GetCleanInputsWithinInterval(level, begin_key, end_key, &files,
+                                             -1 /* hint_index */,
+                                             nullptr /* file_index */);
+      for (const auto& file_meta : files) {
+        if (file_meta->being_compacted) {
+          continue;
+        }
+        if (!include_end && end != nullptr &&
+            cfd->user_comparator()->Compare(file_meta->largest.user_key(),
```

Comment:
should this check be against file smallest key? i.e `Compare(file_meta->smallest.user_key(), *end) >= 0`.

Reply:
seems no need, I think it is a promise of `GetCleanInputsWithinInterval`. The check here is to consider `include_end`.

Reply:
ok, seems this is the same as `rocksdb::DeleteFilesInRange`